Realm/Realm/Native/SynchronizationContextScheduler.cs

```diff
@@ -0,0 +1,150 @@
+﻿////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using Realms.Native;
+
+namespace Realms
+{
+    internal static class SynchronizationContextScheduler
+    {
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate IntPtr get_context();
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate void post_on_context(IntPtr context, ContextPostHandler callback, IntPtr user_data);
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate void release_context(IntPtr context);
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        [return: MarshalAs(UnmanagedType.U1)]
+        private delegate bool is_on_context(IntPtr context, IntPtr targetContext);
+
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate void ContextPostHandler(IntPtr user_data);
+
+        [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_install_scheduler_callbacks", CallingConvention = CallingConvention.Cdecl)]
+        private static extern void install_scheduler_callbacks(get_context get, post_on_context post, release_context release, is_on_context is_on);
+
+        private class Scheduler
+        {
+            internal SynchronizationContext Context { get; private set; }
+
+            private volatile bool _isReleased;
+
+            internal Scheduler(SynchronizationContext context)
+            {
+                Debug.Assert(context != null, "The SynchronizationContext Scheduler implementation always needs a SynchronizationContext");
+                Context = context;
+            }
+
+            internal void Post(ContextPostHandler callback, IntPtr user_data)
```

Comment:
Isn't this a bit roundabout? The way I'm reading the code, the native schedule invokes this function with a function to be called (the static `handle` function) and an argument another function (`user_data`), that `handle` will invoke? Since `handle` is static, can't we expose it as a C function and always call that instead of passing a pointer to `handle` on every `Post` invocation?